<a href="https://colab.research.google.com/github/C-Fer/Making-predictions-with-neural-networks/blob/main/Making_predictions_with_neural_networks.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Making-predictions-with-neural-networks
# INTELIGENCIA ARTIFICIAL
# Redes neuronales artificiales

## Integrantes:

    Camilo Andrés Fernández Echavez    T00054759
    Edgar David Hernández Medina       T00054326
    René Arturo De Oro Larios          T00054355

## Importación de librerias necesarias

In [ ]:
#Gestión de datos
import os
import pandas as pd
import numpy as np
%matplotlib inline
import datetime

#Librerias para la creación de la red neuronal
from keras.models import Sequential
from keras.layers import Dense, Dropout
from keras.wrappers.scikit_learn import KerasClassifier
from keras.utils import np_utils
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from numpy import  argmax

## Lectura de datos

In [ ]:
# Lectura del archivo CSV filtrado
DataFrame = pd.read_csv("clean_data.csv")
DataFrame.head()

Unnamed: 0  Edad  ... Resultado  CodigoDiagnosticoPrincipal
0           1    46  ...      True                        I10X
1           2    68  ...      True                        I10X
2           3    53  ...      True                        I10X
3           4    88  ...      True                        I10X
4           5    63  ...      True                        I10X

[5 rows x 50 columns]

##Label Encoder

In [ ]:
cdp_types = ('I10X','E119','E109','E149','E785','Z000','I10I','I119','E148','E108','E121','A061','I498','E106','I059','A219','G409','I158','A030','E871','L681','I120','E118','A078')
cdp_df = pd.DataFrame(cdp_types, columns=['cdp_types'])
labelencoder = LabelEncoder()
cdp_df['CodigoDiagnosticoPrincipal'] = labelencoder.fit_transform(cdp_df['cdp_types'])

rsss_types = ('Subsidiado','Contributivo','Especial','No afiliado')
rsss_df = pd.DataFrame(rsss_types, columns=['rsss_types'])
# creating instance of labelencoder
labelencoder = LabelEncoder()
rsss_df['RSSS'] = labelencoder.fit_transform(rsss_df['rsss_types'])

DataFrame = DataFrame.drop(['DiagnosticoPrincipal'],axis=1)

## Se reemplazan los valores categóricos por valores numéricos

In [ ]:
DataFrame = DataFrame.replace({'I10X':17,'E119':8,'E109':6,'E149':11,'E785':12,'Z000':23,'I10I':16,'I119':18,'E148':10,'E108':5,'E121':9,'A061':1,'I498':21,'E106':4,'I059':15,'A219':3,'G409':14,'I158':20,'A030':0,'E871':13,'L681':22,'I120':19,'E118':7,'A078':2})
DataFrame = DataFrame.replace({True: 1,False:0})
DataFrame = DataFrame.replace({'Subsidiado':3,'Contributivo': 0, 'Especial':1,'No afiliado':2})


In [ ]:
DataFrame.head()

Unnamed: 0  Edad  ...  Resultado  CodigoDiagnosticoPrincipal
0           1    46  ...          1                          17
1           2    68  ...          1                          17
2           3    53  ...          1                          17
3           4    88  ...          1                          17
4           5    63  ...          1                          17

[5 rows x 49 columns]

In [ ]:
numero_clases= Y['CodigoDiagnosticoPrincipal'].unique()
print(len(numero_clases))

24


## Definimos las variables de caracteristicas y la variable objetivo

In [ ]:
Y = DataFrame[['CodigoDiagnosticoPrincipal']]
X = DataFrame[[x for x in DataFrame if x not in Y]]

In [ ]:
model = Sequential()
model.add(Dense(128, input_dim=48, activation='relu'))
model.add(Dropout(0.2))
model.add(Dense(24, activation='softmax'))

In [ ]:
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [ ]:
encoded_y = labelencoder.fit_transform(Y)
cat_y = np_utils.to_categorical(encoded_y)

/usr/local/lib/python3.7/dist-packages/sklearn/preprocessing/_label.py:251: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


## Dividiendo datos Conjunto de entrenamiento / Conjunto de Pruebas

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, cat_y, test_size=0.25, random_state=0)
print('Conjunto de entrenamiento: ', X_train.shape, y_train.shape)
print('Conjunto de prueba: ', X_test.shape, y_test.shape)

Conjunto de entrenamiento:  (5589, 48) (5589, 24)
Conjunto de prueba:  (1864, 48) (1864, 24)


## Entrenando la red neuronal

In [ ]:
model.fit(X_train, y_train, epochs=300)

Epoch 1/300
175/175 [==============================] - 11s 2ms/step - loss: 569.9261 - accuracy: 0.6906
Epoch 2/300
175/175 [==============================] - 0s 2ms/step - loss: 170.3244 - accuracy: 0.8136
Epoch 3/300
175/175 [==============================] - 0s 2ms/step - loss: 112.1105 - accuracy: 0.8190
Epoch 4/300
175/175 [==============================] - 0s 2ms/step - loss: 80.0508 - accuracy: 0.7979
Epoch 5/300
175/175 [==============================] - 0s 2ms/step - loss: 41.9338 - accuracy: 0.8196
Epoch 6/300
175/175 [==============================] - 0s 2ms/step - loss: 26.4015 - accuracy: 0.8113
Epoch 7/300
175/175 [==============================] - 0s 2ms/step - loss: 18.6604 - accuracy: 0.8109
Epoch 8/300
175/175 [==============================] - 0s 2ms/step - loss: 12.0170 - accuracy: 0.8021
Epoch 9/300
175/175 [==============================] - 0s 2ms/step - loss: 5.6730 - accuracy: 0.8207
Epoch 10/300
175/175 [==============================] - 0s 2ms/step - loss: 3.3

## Perdida y Precisión del modelo

In [ ]:
test_loss, test_acc = model.evaluate(X_test, y_test)

print('Perdida de prueba: ', test_loss)
print('Precisión de prueba (Accuracy): ', test_acc)

59/59 [==============================] - 0s 1ms/step - loss: 0.5288 - accuracy: 0.8916
Perdida de prueba:  0.5288491249084473
Precisión de prueba (Accuracy):  0.8916308879852295


Se obtuvo una perdida del 52% y una precisión del 89%